examples/dreamer/dreamer.py

```diff
@@ -0,0 +1,355 @@
+        frame_skip=cfg.frame_skip,
+        policy_exploration=policy,
+        recorder=recorder,
+        record_interval=cfg.record_interval,
+        log_keys=cfg.recorder_log_keys,
+    )
+
+    final_seed = collector.set_seed(cfg.seed)
+    print(f"init seed: {cfg.seed}, final seed: {final_seed}")
+    ## Training loop
+    collected_frames = 0
+    pbar = tqdm.tqdm(total=cfg.total_frames)
+    r0 = None
+    path = Path("./log")
+    path.mkdir(exist_ok=True)
+
+    scaler = GradScaler()
+    torch.cuda.empty_cache()
+    for i, tensordict in enumerate(collector):
+
+        # update weights of the inference policy
+        collector.update_policy_weights_()
+
+        if r0 is None:
+            r0 = tensordict["reward"].mean().item()
+        pbar.update(tensordict.numel())
+        if "mask" in tensordict.keys():
+            current_frames = tensordict["mask"].sum()
+            tensordict = tensordict[tensordict.get("mask").squeeze(-1)]
+        else:
+            current_frames = tensordict.numel()
+            collected_frames += current_frames
+        tensordict = tensordict.reshape(-1, cfg.batch_length)
+        replay_buffer.extend(tensordict.cpu())
+
+        if collected_frames >= cfg.init_random_frames:
+            for j in range(cfg.optim_steps_per_batch):
+                # sample from replay buffer
+                sampled_tensordict = replay_buffer.sample(cfg.batch_size).to(device)
+
+                with autocast(dtype=torch.float16):
+                    model_loss_td, sampled_tensordict = world_model_loss(
+                        sampled_tensordict
+                    )
+                    if cfg.record_video:
+                        world_model_td = sampled_tensordict.clone().select(
+                            "pixels", "reco_pixels", "posterior_states", "next_belief"
+
+                        )[:4].detach()
+                    actor_loss_td, sampled_tensordict = actor_loss(sampled_tensordict)
+                    value_loss_td, sampled_tensordict = value_loss(sampled_tensordict)
+
+                scaler.scale(model_loss_td["loss_world_model"]).backward()
+                scaler.scale(actor_loss_td["loss_actor"]).backward()
+                scaler.scale(value_loss_td["loss_value"]).backward()
+
+                scaler.unscale_(world_model_opt)
+                clip_grad_norm_(world_model.parameters(), cfg.grad_clip)
+                scaler.unscale_(actor_opt)
+                clip_grad_norm_(actor_model.parameters(), cfg.grad_clip)
+                scaler.unscale_(value_opt)
+                clip_grad_norm_(value_model.parameters(), cfg.grad_clip)
+
+                scaler.step(world_model_opt)
+                world_model_opt.zero_grad()
+
+                scaler.step(actor_opt)
+                actor_opt.zero_grad()
+
+                scaler.step(value_opt)
+                value_opt.zero_grad()
+
+                scaler.update()
+
+                with torch.no_grad(), set_exploration_mode("mode"):
```

Comment:
I personally prefer to have this out of the inner training loop. The reason is that it's easier to control the number of collection steps than the number of training steps, and small changes in the config can have a great impact in the number of evaluation data collections, which has a great physical memory cost and compute time cost.